src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509chain.c

```diff
@@ -0,0 +1,598 @@
+        ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+        // Set any specific options
+        if (loc[options] != NULL)
+        {
+            // checker.setOptions(options);
+            (*env)->CallVoidMethod(env, loc[checker], g_PKIXRevocationCheckerSetOptions, loc[options]);
+        }
+
+        // params.addCertPathChecker(checker);
+        (*env)->CallVoidMethod(env, params, g_PKIXBuilderParametersAddCertPathChecker, loc[checker]);
+    }
+
+    // params.setRevocationEnabled(true);
+    // PKIXCertPathValidatorResult result = (PKIXCertPathValidatorResult)validator.validate(certPathToUse, params);
+    (*env)->CallVoidMethod(env, params, g_PKIXBuilderParametersSetRevocationEnabled, true);
+    loc[result] = (*env)->CallObjectMethod(env, validator, g_CertPathValidatorValidate, certPathToUse, params);
+    if (TryGetJNIException(env, &loc[ex], false /*printException*/))
+    {
+        if (hadErrorsWithoutRevocation)
+        {
+            // Failed without revocation checking - we don't know for sure if these errors were from revocation checking
+            (*env)->CallBooleanMethod(env, ctx->errorList, g_ArrayListAdd, loc[ex]);
+        }
+        else
+        {
+            // Succeeded without revocation checking - errors must be from revocation checking
+            if (ctx->revocationErrorList == NULL)
+            {
+                ctx->revocationErrorList = ToGRef(env, (*env)->NewObject(env, g_ArrayListClass, g_ArrayListCtor));
+            }
+
+            (*env)->CallBooleanMethod(env, ctx->revocationErrorList, g_ArrayListAdd, loc[ex]);
+        }
+    }
+
+    ret = SUCCESS;
+
+cleanup:
+    RELEASE_LOCALS(loc, env);
+    return ret;
+}
+
+int32_t AndroidCryptoNative_X509ChainValidate(X509ChainContext* ctx,
+                                              PAL_X509RevocationMode revocationMode,
+                                              PAL_X509RevocationFlag revocationFlag)
+{
+    assert(ctx != NULL);
+    JNIEnv* env = GetJNIEnv();
+
+    int32_t ret = FAIL;
+    INIT_LOCALS(loc, validatorType, validator, result, ex);
+
+    // String validatorType = "PKIX";
+    // CertPathValidator validator = CertPathValidator.getInstance(validatorType);
+    loc[validatorType] = JSTRING("PKIX");
+    loc[validator] = (*env)->CallStaticObjectMethod(
+        env, g_CertPathValidatorClass, g_CertPathValidatorGetInstance, loc[validatorType]);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+    // PKIXCertPathValidatorResult result = (PKIXCertPathValidatorResult)validator.validate(certPath, params);
+    loc[result] =
+        (*env)->CallObjectMethod(env, loc[validator], g_CertPathValidatorValidate, ctx->certPath, ctx->params);
+    bool hadErrorsWithoutRevocation = TryGetJNIException(env, &loc[ex], false /*printException*/);
+    if (hadErrorsWithoutRevocation)
+    {
+        (*env)->CallBooleanMethod(env, ctx->errorList, g_ArrayListAdd, loc[ex]);
+    }
+
+    if (revocationMode != X509RevocationMode_NoCheck)
```

Comment:
If there were errors without revocation is there any chance of success with revocation? (My gut says no)

If there isn't, then it seems like the revocationMode check should be `else if`, not `if`, and then the redundant error code copy inside `ValidateWithRevocation` can be removed.

But I could just be missing something :smile:.

Reply:
No chance, indeed. Will avoid `ValidateWithRevocation` if there were errors and put `RevocationStatusUnknown` on everything if checking was requested.